web/src/components/njwds-extended/FileInput.tsx

```diff
@@ -0,0 +1,144 @@
+  const fileInputRef = useRef<HTMLInputElement>(null);
+  const BYTES_IN_A_MB = 1048576;
+  const size = props.maxSizeInMegabytes ? props.maxSizeInMegabytes * BYTES_IN_A_MB : undefined;
+
+  const [hasFileBeenRead, setHasFileBeenRead] = useState<boolean>(false);
+  const [fileSizeError, setFileSizeError] = useState({ hasError: false, errorMessage: "" });
+
+  const createAcceptedFileString = (): string | undefined => {
+    if (!props.accept) return undefined;
+    return props.accept.map((it) => FileTypeToInputString[it]).join(",");
+  };
+
+  const onChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+    if (e.target.files && e.target.files.length > 0) {
+      const uploadedFile = e.target.files[0];
+      const reader = new FileReader();
+      // eslint-disable-next-line unicorn/prefer-add-event-listener
+      reader.onload = function (event: ProgressEvent<FileReader>): void {
+        if (!event.target?.result) {
+          return;
+        }
+
+        setHasFileBeenRead(true);
+        props.onChange({
+          base64Contents: (event.target.result as string).split(",")[1],
+          sizeInBytes: uploadedFile.size,
+          fileType: InputStringToFileType[uploadedFile.type],
+          filename: uploadedFile.name,
+        });
+        setFileSizeError({ hasError: false, errorMessage: "" });
+      };
+      reader.readAsDataURL(uploadedFile);
+    }
+  };
+
+  useLayoutEffect(() => {
+    const fileInputElement = fileInputRef.current;
+
+    // Prevent the user from selecting a file that is too large or of the wrong file type
+    const preventInvalidFileSelection = (e: Event): void => {
+      const eventTarget = e.target as HTMLInputElement;
+      if (eventTarget.files && eventTarget.files.length > 0) {
+        const selectedFile = eventTarget.files[0];
+        if (props.accept && !props.accept.includes(InputStringToFileType[selectedFile.type])) {
+          e.stopImmediatePropagation();
+          setFileSizeError({
+            hasError: true,
+            errorMessage: `${selectedFile.name} file format is not supported.`,
+          });
+        } else if (size && selectedFile.size > size) {
+          e.stopImmediatePropagation();
+          setFileSizeError({
+            hasError: true,
+            errorMessage: `${selectedFile.name} exceeds maximum size of ${props.maxSizeInMegabytes} MB.`,
+          });
+        }
+      }
+    };
+
+    // Event listener is added to the file input element ahead of USWDS listeners - React onChange will be added after
+    fileInputElement?.addEventListener("change", (e: Event) => {
+      preventInvalidFileSelection(e);
+    });
+    if (typeof fileInput.on === "function" && typeof fileInput.off === "function") {
+      fileInput.on(fileInputElement); // initialize USWDS fileInput component
+    }
+  }, [fileInput, props.accept, size, props.maxSizeInMegabytes]);
+
+  useEffect(() => {
```

Comment:
because this is a weird useEffect, it might be nice to use the named-function approach to a useEffect here, so that we can give it a function name to help make it more readable & explain what's going on for people reading it

Reply:
included